magma/lib/magma/attributes/file_collection_attribute.rb

```diff
@@ -0,0 +1,155 @@
+require_relative './file_copier'
+require_relative './file_serializer'
+
+class Magma
+  class FileCollectionAttribute < Attribute
+    def database_type
+      :json
+    end
+
+    def serializer
+      @serializer ||= FileSerializer.new(magma_model: @magma_model, attribute: self)
+    end
+
+    def revision_to_loader(record_name, files)
+      loader_format = files.each_with_index.map do |revision, index|
+        serializer.to_loader_format(record_name, revision, index)
+      end
+      [ name, loader_format ]
+    end
+
+    def revision_to_payload(record_name, files, loader)
+      payload_format = files.each_with_index.map do |file_hash, index|
+        serializer.to_payload_format(record_name, file_hash, loader.user, index)
+      end
+      [name, payload_format ]
+    end
+
+    def query_to_payload(data)
+      return nil unless data
+
+      data.map do |datum|
+        serializer.to_query_payload_format(datum)
+      end
+    end
+
+    def query_to_tsv(files)
+      files.map do |file|
+        serializer.to_query_tsv_format(file)
+      end
+    end
+
+    def entry(files, loader)
+      entry_format = files.map do |file|
+        serializer.to_loader_entry_format(file)
+      end.compact  # Remove the ::temp `nil` values for the loader
+
+      [ column_name, entry_format.to_json ]
+    end
+
+    def load_hook(loader, record_name, files, copy_revisions)
+      # Revising a FileCollectionAttribute makes some assumptions about the revisions.
+      # We need to make sure that the links aren't lost on the Metis
+      #   end of things, since that's the only pointer to the pre-revision data_block
+      #   that we have for each Magma file.
+      # For example,
+      #     [file[1], file[0]] is a bad revision, because on Metis's end,
+      #   after we update the new index-0 file (magma-file-0 -> original-magma-file-1-data-block),
+      #   we lose the pointer to the original magma-file-0 data_block.
+      #   It becomes impossible to do the magma-file-1 -> original-magma-file-0-data-block
+      #   revision.
+      # Thus, we must enforce these requirements:
+      #   1) Any existing files must appear in ascending index order.
+      #       [file[0], file[1], file[7], file[11]] is okay.
+      #       [file[11], file[1], file[0], file[7]] is NOT okay.
+      #   2) New files / Metis paths must appear at the end of the sequence.
+      #       [file[1], file[8], {path: "metis://foo/bar/bim.txt"}] is okay.
+      #       [{path: "metis://bad/revision/order.txt"}, file[1], file[8]] is NOT okay.
+      #   3) Any ::temp files must be at the end of the array, so the filename indices stay sequential.
```

Comment:
Not sure keeping the `::temp` files at the end is strictly necessary, but I probably won't figure that out until implementing the upload for FileCollection.